molecules/data_extractor/data_extractor.py

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python
+
+from __future__ import print_function
+
+import ftplib
+import multiprocessing as mp
+import os
+import StringIO
+import tensorflow as tf
+import zlib
+
+FTP_SERVER = 'ftp.ncbi.nlm.nih.gov'
+FTP_USER = 'anonymous'
+FTP_PASS = 'guest'
+FTP_PATH = 'pubchem/Compound_3D/01_conf_per_cmpd/SDF'
+
+# Good for debugging
+FORCE_DISABLE_MULTIPROCESSING = False
+
+
+def _function_wrapper(args_tuple):
+  function, args = args_tuple
+  try:
+    return function(*args)
+  except KeyboardInterrupt:
+    pass
+
+
+def parallel_map(function, iterable):
+  if FORCE_DISABLE_MULTIPROCESSING:
+    return [function(*args) for args in iterable]
+
+  num_threads = mp.cpu_count() * 2
+  pool = mp.Pool(processes=num_threads)
+  p = pool.map_async(_function_wrapper, ((function, args) for args in iterable))
+  try:
+    results = p.get()
+  except KeyboardInterrupt:
+    raise
+  pool.close()
+  return results
+
+
+def extract_data_file(i, N, ftp_path, data_dir):
+  basename = os.path.basename(ftp_path)
+  sdf_file = os.path.join(data_dir, os.path.splitext(basename)[0])
+  progress = '[{i:{n}}/{N}]'.format(i=i+1, n=len(str(N)), N=N)
+
+  if not tf.gfile.Exists(sdf_file):
+    print('{} Extracting data file: {}'.format(progress, sdf_file))
+    # The `ftp` object cannot be pickled for multithreading, so we open a
+    # new connection here
```

Comment:
actually if you use [treading](https://docs.python.org/2/library/threading.html) instead of multiprocessing it might not require pickling the connection object.

Reply:
After trying with `multiprocessing.pool.ThreadPool`, I started getting unclear errors. Apparently the FTP protocol allows only one thing to be fetched at a time, so making the handler a shared resource caused multiple threads to fetch at the same time, and thus erroring out.

Looks like every FTP data channel is supposed to be used for a unique transfer, so every client should open one data connection at a time. I'll keep it as it is so every extraction opens a new connection.